pest/src/iterators_private/pairs.rs

```diff
@@ -0,0 +1,108 @@
+// pest. The Elegant Parser
+// Copyright (C) 2017  Dragoș Tiselice
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use std::rc::Rc;
+
+use super::pair::{self, Pair};
+use super::queueable_token::QueueableToken;
+use super::token_iterator::{self, TokenIterator};
+use super::super::inputs_private::Input;
+use super::super::RuleType;
+
+/// A `struct` containing `Pairs`. It is created in [`pest::state`](../fn.state.html).
+pub struct Pairs<R, I: Input> {
+    queue: Rc<Vec<QueueableToken<R>>>,
+    input: Rc<I>,
+    start: usize,
+    end: usize
+}
+
+pub fn new<R: RuleType, I: Input>(
+    queue: Rc<Vec<QueueableToken<R>>>,
+    input: Rc<I>,
+    start: usize,
+    end: usize
+) -> Pairs<R, I> {
+    Pairs {
+        queue: queue,
+        input: input,
+        start: start,
+        end: end
```

Comment:
I believe field init shorthand was released in 1.17 so you can do `Pairs { queue, input, start, end }`